rllib/utils/replay_buffers/episode_replay_buffer.py

```diff
@@ -0,0 +1,508 @@
+            # initial 0.0 one.
+            rewards[B].extend(episode.rewards[episode_ts:])
+            assert len(observations[B]) == len(actions[B]) == len(rewards[B])
+
+            T = min(len(observations[B]), batch_length_T)
+
+            # Set is_last=True.
+            is_last[B][T-1] = True
+            # If episode is terminated and we have reached the end of it, set
+            # is_terminated=True.
+            if episode.is_terminated and T == len(observations[B]):
+                is_terminated[B][T-1] = True
+            # If episode is truncated and we have reached the end of it, set
+            # is_truncated=True.
+            elif episode.is_truncated and T == len(observations[B]):
+                is_truncated[B][T-1] = True
+
+            # We are done with this batch row.
+            if T == batch_length_T:
+                # We may have overfilled this row: Clip trajectory at the end.
+                observations[B] = observations[B][:batch_length_T]
+                actions[B] = actions[B][:batch_length_T]
+                rewards[B] = rewards[B][:batch_length_T]
+                # Start filling the next row.
+                B += 1
+                T = 0
+
+            # Use next sampled episode/ts pair to fill the row.
+            idx_cursor += 1
+
+        # Update our sampled counter.
+        self.sampled_timesteps += batch_size_B * batch_length_T
+
+        # TODO: Return SampleBatch instead of this simpler dict.
+        ret = {
+            "obs": np.array(observations),
+            "actions": np.array(actions),
+            "rewards": np.array(rewards),
+            "is_first": np.array(is_first),
+            "is_last": np.array(is_last),
+            "is_terminated": np.array(is_terminated),
+            "is_truncated": np.array(is_truncated),
+        }
+
+        return ret
+
+    def get_num_episodes(self) -> int:
+        """Returns the number of episodes (completed or truncated) stored in the buffer.
+        """
+        return len(self.episodes)
+
+    def get_num_timesteps(self) -> int:
+        """Returns the number of individual timesteps stored in the buffer."""
+        return len(self._indices)
+
+    def get_sampled_timesteps(self) -> int:
+        """Returns the number of timesteps that have been sampled in buffer's lifetime.
+        """
+        return self.sampled_timesteps
+
+    @override(ReplayBufferInterface)
+    def get_state(self) -> Dict[str, Any]:
+        return {
+            "episodes": [eps.get_state() for eps in self.episodes],
+            "episode_id_to_index": list(self.episode_id_to_index.items()),
+            "_num_episodes_ejected": self._num_episodes_ejected,
+            "_indices": self._indices,
+            "size": self.size,
+            "sampled_timesteps": self.sampled_timesteps,
+        }
+
+    @override(ReplayBufferInterface)
+    def set_state(self, state) -> None:
+        assert state[0][0] == "episodes"
```

Comment:
Why are we not indexing as a dict here?

Reply:
Yeah, great question. My original implementation was using numpy objects (instead of dict). I'll fix this ...

Reply:
done